Remove dependency on six - we're Py3 only now!

Zac-HD opened the pull request:
This should be pretty simple (if tedious) to review - `six` provides very simple translations so each snippet is obvious; there's just a lot of them!

WeatherGod commented:
Previously, I would have said that this is a good idea, but now I am
concerned about this possibly causing problems for backporting to v2.2.x.

On Wed, May 2, 2018 at 10:36 AM, Zac Hatfield-Dodds <
notifications@github.com> wrote:

> This should be pretty simple (if tedious) to review - six provides very
> simple translations so each snippet is obvious; there's just a lot of them!
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/matplotlib/matplotlib/pull/11158
> Commit Summary
>
>    - Remove dependency on "six" - we're Py3 only!
>
> File Changes
>
>    - *M* INSTALL.rst
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-0> (1)
>    - *M* lib/matplotlib/__init__.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-1> (7)
>    - *M* lib/matplotlib/animation.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-2> (2)
>    - *M* lib/matplotlib/artist.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-3>
>    (13)
>    - *M* lib/matplotlib/backend_bases.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-4>
>    (10)
>    - *M* lib/matplotlib/backends/_backend_tk.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-5>
>    (30)
>    - *M* lib/matplotlib/backends/backend_agg.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-6> (2)
>    - *M* lib/matplotlib/backends/backend_cairo.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-7>
>    (12)
>    - *M* lib/matplotlib/backends/backend_ps.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-8>
>    (26)
>    - *M* lib/matplotlib/backends/backend_svg.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-9>
>    (67)
>    - *M* lib/matplotlib/backends/backend_wx.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-10>
>    (6)
>    - *M* lib/matplotlib/backends/backend_wxagg.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-11>
>    (2)
>    - *M* lib/matplotlib/blocking_input.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-12>
>    (1)
>    - *M* lib/matplotlib/dviread.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-13>
>    (2)
>    - *M* lib/matplotlib/lines.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-14>
>    (10)
>    - *M* lib/matplotlib/mlab.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-15>
>    (33)
>    - *M* lib/matplotlib/patches.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-16>
>    (23)
>    - *M* lib/matplotlib/rcsetup.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-17>
>    (50)
>    - *M* lib/matplotlib/sphinxext/plot_directive.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-18>
>    (47)
>    - *M* lib/matplotlib/tests/__init__.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-19>
>    (2)
>    - *M* lib/matplotlib/tests/test_axes.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-20>
>    (6)
>    - *M* lib/matplotlib/tests/test_backend_svg.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-21>
>    (2)
>    - *M* lib/matplotlib/tests/test_basic.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-22>
>    (10)
>    - *M* lib/matplotlib/tests/test_cbook.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-23>
>    (2)
>    - *M* lib/matplotlib/tests/test_colors.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-24>
>    (7)
>    - *M* lib/matplotlib/tests/test_constrainedlayout.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-25>
>    (1)
>    - *M* lib/matplotlib/tests/test_container.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-26>
>    (1)
>    - *M* lib/matplotlib/tests/test_mathtext.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-27>
>    (2)
>    - *M* lib/matplotlib/tests/test_mlab.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-28>
>    (8)
>    - *M* lib/matplotlib/tests/test_patches.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-29>
>    (6)
>    - *M* lib/matplotlib/tests/test_pickle.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-30>
>    (4)
>    - *M* lib/matplotlib/tests/test_png.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-31>
>    (3)
>    - *M* lib/matplotlib/tests/test_rcparams.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-32>
>    (7)
>    - *M* lib/matplotlib/tests/test_style.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-33>
>    (4)
>    - *M* lib/matplotlib/tests/test_table.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-34>
>    (2)
>    - *M* lib/matplotlib/tests/test_text.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-35>
>    (2)
>    - *M* lib/matplotlib/tests/test_tightlayout.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-36>
>    (1)
>    - *M* lib/matplotlib/tests/test_transforms.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-37>
>    (2)
>    - *M* lib/matplotlib/tests/test_ttconv.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-38>
>    (2)
>    - *M* lib/matplotlib/tests/test_type1font.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-39>
>    (2)
>    - *M* lib/matplotlib/ticker.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-40>
>    (31)
>    - *M* lib/matplotlib/type1font.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-41>
>    (10)
>    - *M* lib/mpl_toolkits/axes_grid1/__init__.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-42>
>    (2)
>    - *M* lib/mpl_toolkits/axes_grid1/anchored_artists.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-43>
>    (2)
>    - *M* lib/mpl_toolkits/axes_grid1/axes_divider.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-44>
>    (3)
>    - *M* lib/mpl_toolkits/axes_grid1/axes_grid.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-45>
>    (2)
>    - *M* lib/mpl_toolkits/axes_grid1/axes_rgb.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-46>
>    (14)
>    - *M* lib/mpl_toolkits/axes_grid1/axes_size.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-47>
>    (4)
>    - *M* lib/mpl_toolkits/axes_grid1/colorbar.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-48>
>    (9)
>    - *M* lib/mpl_toolkits/axes_grid1/inset_locator.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-49>
>    (1)
>    - *M* lib/mpl_toolkits/axes_grid1/mpl_axes.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-50>
>    (5)
>    - *M* lib/mpl_toolkits/axes_grid1/parasite_axes.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-51>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/__init__.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-52>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/angle_helper.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-53>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/axis_artist.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-54>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/axisline_style.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-55>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/axislines.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-56>
>    (4)
>    - *M* lib/mpl_toolkits/axisartist/clip_path.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-57>
>    (3)
>    - *M* lib/mpl_toolkits/axisartist/floating_axes.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-58>
>    (3)
>    - *M* lib/mpl_toolkits/axisartist/grid_finder.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-59>
>    (2)
>    - *M* lib/mpl_toolkits/axisartist/grid_helper_curvelinear.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-60>
>    (3)
>    - *M* lib/mpl_toolkits/mplot3d/__init__.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-61>
>    (2)
>    - *M* lib/mpl_toolkits/mplot3d/axes3d.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-62>
>    (6)
>    - *M* lib/mpl_toolkits/mplot3d/axis3d.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-63>
>    (2)
>    - *M* lib/mpl_toolkits/mplot3d/proj3d.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-64>
>    (3)
>    - *M* lib/mpl_toolkits/tests/test_axes_grid1.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-65>
>    (2)
>    - *M* lib/mpl_toolkits/tests/test_axisartist_clip_path.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-66>
>    (4)
>    - *M* setupext.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-67>
>    (1)
>    - *M* tools/gh_api.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-68>
>    (4)
>    - *M* tools/make_icons.py
>    <https://github.com/matplotlib/matplotlib/pull/11158/files#diff-69>
>    (4)
>
> Patch Links:
>
>    - https://github.com/matplotlib/matplotlib/pull/11158.patch
>    - https://github.com/matplotlib/matplotlib/pull/11158.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/11158>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-AbPc5NLs562b8kS6P6PCEHaF-CAks5tucRjgaJpZM4Tvk2D>
> .
>

jklymak commented:
We discussed a bit on the weekly call.  

I'm a little hazy on the exact implementation, but I think that the issue here is not the PR per se, which is against 3.0, but if someone writes code that is meant to be back ported as well as going against master they will not write python-2 compatible code if all the six stuff is taken out.  Other devs piped up that they aren't dropping harmless py2 stuff until they end long-term-support of 2.0.  

I'm not experienced enough with transitions like this to know whether this is all manageable, but my personal vote would be to hold off on this sort of change until we have it all sorted out.

Zac-HD commented:
Fair enough!

jklymak commented:
Well, I didn't mean for you to close it yet - I was hoping someone more experienced would weigh in...  I'll re-open just because we really should understand the py3 migration path...

anntzer commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -340,14 +334,6 @@ def split_code_at_show(text):
     return parts
 
-
-def remove_coding(text):
```
deprecation, yada yada

anntzer replied:
Far too many things that should not be public API unfortunately are.

jklymak commented:
I’m fine w that.  But what’s the path if I need to fix a regression in 2.2.3?   Code versus master and then fix the backport manually adding back in any six code needed?

WeatherGod commented:
My main concern is about changes like these is that it will make it more
likely for other backports to encounter conflicts, which makes it more work
for us to perform backports. And it isn't like six is keeping us from
taking advantage of py3k features. We are still going ahead with kwarg-only
arguments (although, I am a bit wary of that causing merge conflicts, too).
But usage of six is all over the codebase.

On Wed, May 2, 2018 at 7:27 PM, Jody Klymak <notifications@github.com>
wrote:

> I’m fine w that. But what’s the path if I need to fix a regression in
> 2.2.3? Code versus master and then fix the backport manually adding back in
> any six code needed?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/11158#issuecomment-386151753>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-OvhBWXyMbsUr9jpTgnftc7t3pOqks5tukDPgaJpZM4Tvk2D>
> .
>

ImportanceOfBeingErnest commented:
If I recall correctly the main argument for dropping py2 support is to relieve strain from developpers so they can spend their time on implementing new features instead of working around incompatibility issues.  
Now it seems people spend more time disassembling a previously working code, finding arguments why that would be useful and fighting with the issues brought up by the need to keep diverging branches in sync, instead of using the freedom of py3 to fix existing problems (which are hard to come by with py2) or implementing new features (without the need to think about backwards compatibility). 
If the strategy here is to make backporting as hard as possible such that even the simplest documentation fixes cannot be brought to the 2.2 branch any more and the meaning of "critical bug" becomes closer and closer to "complete failure" the whole "LTS up to 2020" is just hypocritical.

efiring commented:
I'm also in favor of stripping out six now.  It improves readability.  I doubt it will substantially impede critical backports.  (Many of the uses of six--many instances of 'six.iteritems' for example--are not even necessary for compatibility.)

tacaswell commented:
Again, I am not super worried about backport because they should only be for critical bugs and regressions.  If the backports are getting harder (and they are going to get harder independent if we strip six out or not) we then we do less non-critical backports.

> I think the real question is the following: Do you expect new PRs to use six and/or to be Py2-compatible? If yes, then we haven't really gained anything. If no, then someone will have to do the work regarding backporting in any case...

I would say no, we should not be making any concessions to py2 compatibility in new code.

jklymak commented:
Thanks @efiring, @WeatherGod , @tacaswell  @anntzer  and @ImportanceOfBeingErnest  I think that clarifies things for me a fair bit; to paraphrase:

1. build for 3.0, even critical bug fixes. 
2. be prepared to do a bit of fiddling during the backport.  ...and/or ask for help.
3. we should not let the backporting pain slow down adding py3 improvements (like this PR).

Zac-HD commented:
Thanks for all your comments - I will keep working on this as the majority opinion in in favor.

My main motivation is to remove a layer of unnecessary complexity - contributing to Matplotlib (or just reading the code to debug a plot) sometimes feels like death by a thousand papercuts.  Removing `six` is a small step in the right direction, and if we keep making incremental progress it adds up to qualitative change!

ImportanceOfBeingErnest commented:
Mind that this is not only about code, it is also about documentation. An error in the docs can be as critical as an error in the code from a user's perspective. I recently submitted a PR to fix the documentation, which currently has a major part of it completely undocumented (the axes_grid1 and axisartist toolkits). This could not be backported to the current doc branch because some `future` imports conflicting - totally unrelated to the actual PR. 
The problem is for such bug fixes that in principle they would be easily solved by the person doing the PR, yet due to the backporting it is a task that only a handful of people are capable of doing. 
Rightfully they refuse to do that as it is more work for them, which they should better spend on reviewing all the other open PR etc. 

Ideally there should be the option to pull/push from/to the p2 branch independently with a clear manual on how to do that. There should be ci testing available to make sure fixes don't break anything. You can then leave it to the people doing the PRs to make sure everything runs smoothly. On a by-case basis critical fixes can be brought to that branch by the dev team, but essentially you have the freedom to completely transfer responsibility (appart from reviewing) to the contributors. The question whether or not a bug is critical is then simply a question of whether someone is willing to do that extra work (which usually isn't that much anyways).

Zac-HD commented:
Surely anyone could `got cherry-pick` the commits, fix merge conflicts, and open a PR against the 2.2 branch though? I don't see what makes this harder than any other contribution.

Backports aren't automatic, and taking six out doesn't make them harder.

WeatherGod commented:
Actually, we do have an automation system for backports that merge cleanly.
It is pretty awesome.

On Thu, May 3, 2018 at 9:09 AM, Zac Hatfield-Dodds <notifications@github.com
> wrote:

> Surely anyone could got cherry-pick the commits, fix merge conflicts, and
> open a PR against the 2.2 branch though? I don't see what makes this harder
> than any other contribution.
>
> Backports aren't automatic, and taking six out doesn't make them harder.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/11158#issuecomment-386289772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-PbUwmKX2LN3rRqEXB-utTa8GlRWks5tuwGYgaJpZM4Tvk2D>
> .
>

Zac-HD commented:
I've seen that and it's great, but a clean merge is not always a correct merge - there should still be a human in the loop!

We're getting off topic though - I'd appreciate it if we could stick at least mostly to review of code. I'll push an update tomorrow with a first round of fixes...

jklymak commented:
I think a PR that changes 70 files is an appropriate place to discuss the practical implications of the PR for back porting.

tacaswell commented:
> Mind that this is not only about code, it is also about documentation. An error in the docs can be as critical as an error in the code from a user's perspective. I recently submitted a PR to fix the documentation, which currently has a major part of it completely undocumented (the axes_grid1 and axisartist toolkits). This could not be backported to the current doc branch because some future imports conflicting - totally unrelated to the actual PR.

I agree errors in documentation as as or more important than bugs, however we should apply the same criteria to documentation as we do to code.  Those modules have never been documented so adding them is not fixing a regression from 2.x series.

> The problem is for such bug fixes that in principle they would be easily solved by the person doing the PR, yet due to the backporting it is a task that only a handful of people are capable of doing.

We should broaden the number of people who feel comfortable doing manual backports then.  We have some minimal instructions at https://matplotlib.org/devel/coding_guide.html#manual-backports What do we need to add to that to make it clearer?



> Ideally there should be the option to pull/push from/to the p2 branch independently with a clear manual on how to do that. There should be ci testing available to make sure fixes don't break anything.

You can checkout and open PRs against 2.2.x and CI is still enabled on that branch.  However, if we start to accumulate lots of development on the v2.2.x branch we will then have to deal with forward porting and backporting will only become more difficult so I am not excited about this. 

> On a by-case basis critical fixes can be brought to that branch by the dev team, but essentially you have the freedom to completely transfer responsibility (appart from reviewing) to the contributors.

Reviewing is where our bottleneck is.

So far no one has volunteered to be the v2.2 release manager.  If someone does I will be happy to defer to what ever they want to do on many of these positions.

dopplershift commented:
> If reviewing is a bottleneck, one smight think about delegating more of that as well. Maybe it can be separated from the rights to merge?

There's nothing stopping non-committers from writing reviews, it works just fine today. They just don't appear as the green check mark.

I'm really not buying this argument that **BUG FIXES**, specifically regressions from >=2.0, are going to have such a large surface area that the removal of six is going to create this huge burden for back ports. Honestly, this feels like trepidation at the dropping of 2.7 (from those who still have to use it) rather than a solid technical concern.

WeatherGod commented:
> Honestly, this feels like trepidation at the dropping of 2.7 (from those
who still have to use it) rather than a solid technical concern.

This has *nothing* to do with trepidations at the dropping of 2.7. We are
already well-committed down that path with the changes in master as it
stands. I have put up *zero* resistance to the transistions to keyword-only
arguments and such (even gave a +1 in another issue to rework a bit of
mplot3d code to facilitate that).

Unlike changes to our imports and changes to the function call signatures,
removing six is a change that is spread through-out the codebase. The
chances for a backport conflict increases, which makes for extra work that
would not have been needed if six is left in the codebase right now. Also,
as noted before, it becomes more likely that we will accidentally backport
broken py27 code if we don't code using six for the near-term future. Also,
not removing six does not impede adoption of py3k features.

The only benefit I have seen suggested is improved readability. Quite
frankly, I have gotten so used to six the past few years, that I really
don't notice it anymore. Now, compare that against increased work for
backports and accidental introduction of subtle bugs for what is supposed
to be bugfix versions, I really don't see the value proposition here.

As for release manager for v2.2.x, I think we all know that I was going to
take that role, I was just hoping someone else would step up instead...


On Thu, May 3, 2018 at 5:32 PM, Ryan May <notifications@github.com> wrote:

> If reviewing is a bottleneck, one smight think about delegating more of
> that as well. Maybe it can be separated from the rights to merge?
>
> There's nothing stopping non-committers from writing reviews, it works
> just fine today. They just don't appear as the green check mark.
>
> I'm really not buying this argument that *BUG FIXES*, specifically
> regressions from >=2.0, are going to have such a large surface area that
> the removal of six is going to create this huge burden for back ports.
> Honestly, this feels like trepidation at the dropping of 2.7 (from those
> who still have to use it) rather than a solid technical concern.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/11158#issuecomment-386443618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-AEjNjnqP5KvRIvN68JGT1x5QfCyks5tu3d9gaJpZM4Tvk2D>
> .
>

efiring commented:
On 2018/05/03 11:47 AM, Benjamin Root wrote:
> Also,
> as noted before, it becomes more likely that we will accidentally backport
> broken py27 code if we don't code using six for the near-term future.

To me, this idea of continuing to code using six on master makes no 
sense at all, in the context of master as a py3-only codebase--which it 
already is, as was planned.  I don't think developers on master should 
be burdened by having to keep track of whether six would be needed if 
the code were backported.

As for breakage: the backports will still be tested against py27.

If there is a backport conflict related to six, it should be pretty 
obvious and easy to fix.

There should not be a large number of backports if we stick to Thomas's 
criteria.

Zac-HD commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -540,7 +538,7 @@ def show(self):
         PyEval_RestoreThread: NULL state bug on win32
         """
-        _focus = windowing.FocusManager()
+        windowing.FocusManager()
```
Oh... dear.  It's actually not used in many places, but I'll convert it in a seperate pull.

tacaswell commented on lib/matplotlib/backends/backend_cairo.py:
```diff
@@ -275,11 +273,7 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
-            else:
-                if six.PY2 and isinstance(s, six.text_type):
-                    s = s.encode("utf-8")
+                s = str(s)
```
attn @anntzer 

This logic does not seem identical to be, but I am not sure of the context.

Zac-HD replied:
The `else` block is a no-op on Python 3, so it can be removed.  That leaves the former block, and we can simplify the condition by observing that `str(some_string)` is a no-op too, and therefore calling `s = str(s)` unconditionally.

anntzer replied:
Pretty sure the call to str/six.text_types was only necessary because of unicode/str issues; now *s* should always be a str and we shoudn't even have to call `str(s)`.

tacaswell commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1,9 +1,6 @@
 A PostScript backend, which can produce both PostScript .ps and .eps
 """
-import six
-from six.moves import StringIO
```
Do we not use `StringIO` in this module?

Zac-HD replied:
Not anymore - it was only used in a `isinstance` check in a Python-2-only branch.

tacaswell commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -981,8 +975,7 @@ def _print_figure(
         isEPSF = format == 'eps'
-        if isinstance(outfile,
-                      (six.string_types, getattr(os, "PathLike", ()),)):
+        if isinstance(outfile, (str, getattr(os, "PathLike", ()),)):
```
This `getattr` looks like another back-compat shim, how far back do we need this?

Zac-HD replied:
`os.PathLike` is new in Python 3.6, so it will be needed for a while yet.

tacaswell commented:
Only made it halfway through, but quite surprised at how _small_ this diff is.

Zac-HD commented:
Ping @tacaswell - I've rebased on master to resolve a merge conflict.  Would you mind reviewing the other half of the diff so we can merge this before another conflict comes up?

QuLogic commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -943,12 +941,12 @@ def trigger(self, *args):
                 # save dir for next time
                 rcParams['savefig.directory'] = os.path.dirname(
-                    six.text_type(fname))
+                    str(fname))
```
Should fit on one line now?

QuLogic commented on lib/matplotlib/patches.py:
```diff
@@ -1846,24 +1843,12 @@ def _pprint_styles(_styles):
-                    for argname, argdefault in zip(args[1:], defaults)]
+        spec = inspect.getfullargspec(cls.__init__)
+        if spec.defaults:
+            argstr = ",".join("%s=%s" % pair for pair in zip(
```
I would still break at the `for` here (and the `,` if necessary).

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1982,15 +1976,6 @@ def decade_up(x, base=10):
     return base ** lx
 
 
-def nearest_long(x):
```
`round` does banker's rounding in Python 3; `nearest_long` does not.

Zac-HD replied:
...yep, that's not `round`.  It's `int`.  /facepalm.

QuLogic replied:
Sorry for being unclear; it's not `int` either, it's whatever the other kind of rounding is.

Zac-HD replied:
Yep, round-half-to-zero.  Fixed by inlining the logic in the two places it was needed, and just using `round` when we know the value is very close to an integer.

anntzer replied:
Also, public api, yada yada.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1120,15 +1114,13 @@ def __call__(self, x, pos=None):
             return self._non_decade_format(sign_string, base, fx, usetex)
+        elif usetex:
+            return r'$%s%s^{%d}$' % (
+                sign_string, base, int(fx + (-.5 if fx < 0 else .5)))
```
actually fx is close to int here (the "not is_x_decade" case is handled just above) so it has already been rounded above, so you don't need to call any of nearest_long, round, or your inlined version.

Zac-HD commented:
Hi @anntzer & @tacaswell - I've rebased on master, split it into separate commits for easier reviewing, and deprecated instead of removing the various potentially-public functions.  Ready for final review!

Please leave a fresh comment if I've missed anything - this has been a long thread and I'm keen to get this merged before I have to rebase *again* :smile:

anntzer commented on lib/matplotlib/tests/test_basic.py:
```diff
@@ -22,14 +21,7 @@ def test_override_builtins():
-    else:
-        builtins = sys.modules['__builtin__']
-
+    builtins = sys.modules['builtins']
```
or just "import builtins"

anntzer commented:
Technically you also need an entry in doc/next_api_changes/ but I'm not going to hold up the PR for that (and likewise I can handle the comment above in a separate PR).

anntzer commented on lib/matplotlib/patches.py:
```diff
@@ -1848,24 +1845,13 @@ def _pprint_styles(_styles):
+        if spec.defaults:
+            argstr = ",".join("%s=%s" % pair
+                              for pair in zip(spec.args[-len(spec.defaults):],
+                                              spec.defaults))
```
would write this as
```
",".join(map("{}={}".format, spec.args[-len(spec.defaults):], spec.defaults))
```
(the rarely seen but ever useful implicit zip-and-star in multi-arg map).

(would also leave a space after the comma, but that's a separate point)

anntzer commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -466,10 +461,7 @@ def run_code(code, code_path, ns=None, function_name=None):
-    else:
-        pwd = os.getcwd()
+    pwd = os.getcwd()
     old_sys_path = list(sys.path)
```
would rewrite as `sys.path.copy()`, otherwise it's not immediately clear what's happening.

anntzer commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -494,10 +486,7 @@ def run_code(code, code_path, ns=None, function_name=None):
-        sys.stdout = io.StringIO()
-    else:
-        sys.stdout = cStringIO.StringIO()
+    sys.stdout = io.StringIO()
```
see if contextlib.redirect_stdout works here

Zac-HD replied:
*probably* would, so I'm leaving it for later.

anntzer commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -512,17 +501,16 @@ def _dummy_print(*arg, **kwarg):
-                six.exec_("__name__ = '__main__'", ns)
-            code = remove_coding(code)
-            six.exec_(code, ns)
+                exec("__name__ = '__main__'", ns)
```
`ns["__name__"] = "__main__"`

Zac-HD commented:
Thanks @anntzer!  I've done the easy bits here :smile: 

I *think* `contextlib.redirect_stdout` would both work and remove the need for a `_dummy_print` function, but it's not obviously correct so I'm leaving that for a future PR as you suggested.  Similarly I don't see what the API change would be, unless either deprecations count or users are accessing imported names (which are non-public, per PEP8).

So... merge please?

jklymak commented:
@Zac-HD  Remember we a) have a really long review queue; this is 10-days old, @anntzer has some things that are 10 months old I'm sure he'd like reviewed as well; b) we require two reviews before things get merged.  So you'll have to get someone else excited about going through this :wink:  Its not going to be me, because my knowledge of py2->py3 is pretty minimal.  @tacaswell started, but is often quite busy, but he seemed interested so I'm sure he'll get to it.

Zac-HD commented:
Ping @jklymak, with many thanks 😄
